File: legacy-node/src-js/services/antiSpamService.js
import { BotError } from '../domain.js';

export class AntiSpamService {
  constructor(storage, config) {
    this.storage = storage;
    this.config = config;
  }

  assertInputLength(text) {
    if ((text || '').length > this.config.maxInputChars) {
      throw new BotError(`Слишком длинный запрос. Лимит: ${this.config.maxInputChars} символов.`, 'TOO_LARGE');
    }
  }

  enforceOutputLength(text) {
    if ((text || '').length <= this.config.maxOutputChars) return text;
    return `${text.slice(0, this.config.maxOutputChars)}…`;
  }

  enforceCooldown(userId, scope = 'ai') {
    const now = Math.floor(Date.now() / 1000);
    const key = `cooldown:${scope}:${userId}`;
    const row = this.storage.getRateLimit(key);
    if (row && Number(row.cooldown_until) > now) {
      const wait = Number(row.cooldown_until) - now;
      throw new BotError(`Слишком часто. Повторите через ${wait} сек.`, 'RATE_LIMITED');
    }
    this.storage.saveRateLimit(key, now, 0, now + this.config.aiCooldownSec);
  }

  bumpWindow(userId, command, maxPerWindow, windowSec) {
    const now = Math.floor(Date.now() / 1000);
    const key = `window:${command}:${userId}`;
    const row = this.storage.getRateLimit(key);
    const start = row ? Number(row.window_start) : now;
    let count = row ? Number(row.count) : 0;

    if (now - start >= windowSec) {
      count = 0;
    }

    count += 1;
    if (count > maxPerWindow) {
      throw new BotError(`Слишком частое использование ${command}. Попробуйте позже.`, 'RATE_LIMITED');
    }

    this.storage.saveRateLimit(key, now - (now - start >= windowSec ? 0 : now - start), count, 0);
  }
}

File: src/main/java/com/dauren/services/AntiSpamService.java
package com.dauren.services;

import com.dauren.storage.SqliteStorage;

public class AntiSpamService {
    private final SqliteStorage storage;
    private final int maxInputChars;
    private final int maxOutputChars;
    private final int aiCooldownSec;

    public AntiSpamService(SqliteStorage storage, int maxInputChars, int maxOutputChars, int aiCooldownSec) {
        this.storage = storage;
        this.maxInputChars = maxInputChars;
        this.maxOutputChars = maxOutputChars;
        this.aiCooldownSec = aiCooldownSec;
    }

    public void assertInputLength(String text) {
        if (text != null && text.length() > maxInputChars) {
            throw new IllegalArgumentException("Слишком длинный запрос.");
        }
    }

    public String trimOutput(String text) {
        if (text == null || text.length() <= maxOutputChars) return text;
        return text.substring(0, maxOutputChars) + "…";
    }

    public void enforceCooldown(String userId, String scope) {
        long now = System.currentTimeMillis() / 1000;
        String key = "cooldown:" + scope + ":" + userId;
        var row = storage.getRate(key).orElse(null);
        if (row != null && row.cooldownUntil() > now) {
            throw new IllegalArgumentException("Слишком часто. Повторите позже.");
        }
        storage.upsertRate(key, now, 0, now + aiCooldownSec);
    }

    public void bumpWindow(String userId, String command, int windowSec, int maxCount) {
        long now = System.currentTimeMillis() / 1000;
        String key = "window:" + command + ":" + userId;
        var row = storage.getRate(key).orElse(null);
        long start = row == null ? now : row.windowStart();
        int count = row == null ? 0 : row.count();
        if (now - start >= windowSec) {
            start = now;
            count = 0;
        }
        count++;
        if (count > maxCount) throw new IllegalArgumentException("Слишком частое использование " + command);
        storage.upsertRate(key, start, count, 0);
    }
}
